core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -94,27 +94,24 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       int columnIndex,
       RelDataType targetType) {
     final SqlKind kind = query.getKind();
+    boolean coerced = false;
     switch (kind) {
     case SELECT:
       SqlSelect selectNode = (SqlSelect) query;
       SqlValidatorScope scope1 = validator.getSelectScope(selectNode);
-      if (!coerceColumnType(scope1, getSelectList(selectNode), columnIndex, targetType)) {
-        return false;
-      }
-      updateInferredColumnType(scope1, query, columnIndex, targetType);
```

Comment:
We are updating the `scope1` instead of `scope`, however, after your change, we'll update `scope`.